File: sample/swift/09_fire_simulation.swift
import Foundation


// 09: Sample that outputs a simple fire effect as a GIF.

func fire_palette() -> [Any] {
    var p: [Any] = __pytra_as_list([])
    let __step_0 = __pytra_int(Int64(1))
    var i = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && i < __pytra_int(Int64(256))) || (__step_0 < 0 && i > __pytra_int(Int64(256)))) {
        var r: Int64 = __pytra_int(Int64(0))
        var g: Int64 = __pytra_int(Int64(0))
        var b: Int64 = __pytra_int(Int64(0))
        if (__pytra_int(i) < __pytra_int(Int64(85))) {
            r = __pytra_int((__pytra_int(i) * __pytra_int(Int64(3))))
            g = __pytra_int(Int64(0))
            b = __pytra_int(Int64(0))
        } else {
            if (__pytra_int(i) < __pytra_int(Int64(170))) {
                r = __pytra_int(Int64(255))
                g = __pytra_int((__pytra_int((__pytra_int(i) - __pytra_int(Int64(85)))) * __pytra_int(Int64(3))))
                b = __pytra_int(Int64(0))
            } else {
                r = __pytra_int(Int64(255))
                g = __pytra_int(Int64(255))
                b = __pytra_int((__pytra_int((__pytra_int(i) - __pytra_int(Int64(170)))) * __pytra_int(Int64(3))))
            }
        }
        p = __pytra_as_list(p); p.append(r)
        p = __pytra_as_list(p); p.append(g)
        p = __pytra_as_list(p); p.append(b)
        i += __step_0
    }
    return __pytra_bytes(p)
}

func run_09_fire_simulation() {
    var w: Int64 = __pytra_int(Int64(380))
    var h: Int64 = __pytra_int(Int64(260))
    var steps: Int64 = __pytra_int(Int64(420))
    var out_path: String = __pytra_str("sample/out/09_fire_simulation.gif")
    var start: Double = __pytra_float(__pytra_perf_counter())
    var heat: [Any] = __pytra_as_list(({ () -> [Any] in var __out: [Any] = []; let __step = __pytra_int(Int64(1)); var __lc_i = __pytra_int(Int64(0)); while ((__step >= 0 && __lc_i < __pytra_int(h)) || (__step < 0 && __lc_i > __pytra_int(h))) { __out.append(__pytra_list_repeat(Int64(0), w)); __lc_i += __step }; return __out })())
    var frames: [Any] = __pytra_as_list([])
    let __step_0 = __pytra_int(Int64(1))
    var t = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && t < __pytra_int(steps)) || (__step_0 < 0 && t > __pytra_int(steps))) {
        let __step_1 = __pytra_int(Int64(1))
        var x = __pytra_int(Int64(0))
        while ((__step_1 >= 0 && x < __pytra_int(w)) || (__step_1 < 0 && x > __pytra_int(w))) {
            var val: Int64 = __pytra_int((__pytra_int(Int64(170)) + __pytra_int((__pytra_int((__pytra_int((__pytra_int(x) * __pytra_int(Int64(13)))) + __pytra_int((__pytra_int(t) * __pytra_int(Int64(17)))))) % __pytra_int(Int64(86))))))
            __pytra_setIndex(__pytra_as_list(__pytra_getIndex(heat, (__pytra_int(h) - __pytra_int(Int64(1))))), x, val)
            x += __step_1
        }
        let __step_2 = __pytra_int(Int64(1))
        var y = __pytra_int(Int64(1))
        while ((__step_2 >= 0 && y < __pytra_int(h)) || (__step_2 < 0 && y > __pytra_int(h))) {
            let __step_3 = __pytra_int(Int64(1))
            var x = __pytra_int(Int64(0))
            while ((__step_3 >= 0 && x < __pytra_int(w)) || (__step_3 < 0 && x > __pytra_int(w))) {
                var a: Int64 = __pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(heat, y)), x)))
                var b: Int64 = __pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(heat, y)), (__pytra_int((__pytra_int((__pytra_int(x) - __pytra_int(Int64(1)))) + __pytra_int(w))) % __pytra_int(w)))))
                var c: Int64 = __pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(heat, y)), (__pytra_int((__pytra_int(x) + __pytra_int(Int64(1)))) % __pytra_int(w)))))
                var d: Int64 = __pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(heat, (__pytra_int((__pytra_int(y) + __pytra_int(Int64(1)))) % __pytra_int(h)))), x)))
                var v: Int64 = __pytra_int((__pytra_int(__pytra_int((__pytra_int((__pytra_int((__pytra_int(a) + __pytra_int(b))) + __pytra_int(c))) + __pytra_int(d))) / __pytra_int(Int64(4)))))
                var cool: Int64 = __pytra_int((__pytra_int(Int64(1)) + __pytra_int((__pytra_int((__pytra_int((__pytra_int(x) + __pytra_int(y))) + __pytra_int(t))) % __pytra_int(Int64(3))))))
                var nv: Int64 = __pytra_int((__pytra_int(v) - __pytra_int(cool)))
                __pytra_setIndex(__pytra_as_list(__pytra_getIndex(heat, (__pytra_int(y) - __pytra_int(Int64(1))))), x, __pytra_ifexp((__pytra_int(nv) > __pytra_int(Int64(0))), nv, Int64(0)))
                x += __step_3
            }
            y += __step_2
        }
        var frame: [Any] = __pytra_as_list(__pytra_bytearray((__pytra_int(w) * __pytra_int(h))))
        let __step_4 = __pytra_int(Int64(1))
        var yy = __pytra_int(Int64(0))
        while ((__step_4 >= 0 && yy < __pytra_int(h)) || (__step_4 < 0 && yy > __pytra_int(h))) {
            var row_base: Int64 = __pytra_int((__pytra_int(yy) * __pytra_int(w)))
            let __step_5 = __pytra_int(Int64(1))
            var xx = __pytra_int(Int64(0))
            while ((__step_5 >= 0 && xx < __pytra_int(w)) || (__step_5 < 0 && xx > __pytra_int(w))) {
                __pytra_setIndex(frame, (__pytra_int(row_base) + __pytra_int(xx)), __pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(heat, yy)), xx)))
                xx += __step_5
            }
            yy += __step_4
        }
        frames = __pytra_as_list(frames); frames.append(__pytra_bytes(frame))
        t += __step_0
    }
    __pytra_noop(out_path, w, h, frames, fire_palette())
    var elapsed: Double = __pytra_float((__pytra_float(__pytra_perf_counter()) - __pytra_float(start)))
    __pytra_print("output:", out_path)
    __pytra_print("frames:", steps)
    __pytra_print("elapsed_sec:", elapsed)
}

@main
struct Main {
    static func main() {
        run_09_fire_simulation()
    }
}

File: sample/go/09_fire_simulation.go
package main


// 09: Sample that outputs a simple fire effect as a GIF.

func fire_palette() []any {
    var p []any = __pytra_as_list([]any{})
    for i := int64(0); i < int64(256); i += 1 {
        var r int64 = int64(0)
        var g int64 = int64(0)
        var b int64 = int64(0)
        if (i < int64(85)) {
            r = (i * int64(3))
            g = int64(0)
            b = int64(0)
        } else {
            if (i < int64(170)) {
                r = int64(255)
                g = ((i - int64(85)) * int64(3))
                b = int64(0)
            } else {
                r = int64(255)
                g = int64(255)
                b = ((i - int64(170)) * int64(3))
            }
        }
        p = append(p, r)
        p = append(p, g)
        p = append(p, b)
    }
    return __pytra_as_list(__pytra_bytes(p))
}

func run_09_fire_simulation() {
    var w int64 = int64(380)
    var h int64 = int64(260)
    var steps int64 = int64(420)
    var out_path string = __pytra_str("sample/out/09_fire_simulation.gif")
    var start float64 = __pytra_perf_counter()
    var heat []any = __pytra_as_list(func() []any { __out := []any{}; __step := __pytra_int(int64(1)); for __lc_i := __pytra_int(int64(0)); (__step >= 0 && __lc_i < __pytra_int(h)) || (__step < 0 && __lc_i > __pytra_int(h)); __lc_i += __step { __out = append(__out, __pytra_list_repeat(int64(0), w)) }; return __out }())
    var frames []any = __pytra_as_list([]any{})
    for t := int64(0); t < steps; t += 1 {
        for x := int64(0); x < w; x += 1 {
            var val int64 = (int64(170) + (((x * int64(13)) + (t * int64(17))) % int64(86)))
            __pytra_set_index(__pytra_as_list(__pytra_get_index(heat, (h - int64(1)))), x, val)
        }
        for y := int64(1); y < h; y += 1 {
            for x := int64(0); x < w; x += 1 {
                var a int64 = __pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(heat, y)), x))
                var b int64 = __pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(heat, y)), (((x - int64(1)) + w) % w)))
                var c int64 = __pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(heat, y)), ((x + int64(1)) % w)))
                var d int64 = __pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(heat, ((y + int64(1)) % h))), x))
                var v int64 = __pytra_int(((((a + b) + c) + d) / int64(4)))
                var cool int64 = (int64(1) + (((x + y) + t) % int64(3)))
                var nv int64 = (v - cool)
                __pytra_set_index(__pytra_as_list(__pytra_get_index(heat, (y - int64(1)))), x, __pytra_ifexp((nv > int64(0)), nv, int64(0)))
            }
        }
        var frame []any = __pytra_as_list(__pytra_bytearray((w * h)))
        for yy := int64(0); yy < h; yy += 1 {
            var row_base int64 = (yy * w)
            for xx := int64(0); xx < w; xx += 1 {
                __pytra_set_index(frame, (row_base + xx), __pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(heat, yy)), xx)))
            }
        }
        frames = append(frames, __pytra_bytes(frame))
    }
    __pytra_save_gif(out_path, w, h, frames, fire_palette(), int64(4), int64(0))
    var elapsed float64 = (__pytra_perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", steps)
    __pytra_print("elapsed_sec:", elapsed)
}

func main() {
    run_09_fire_simulation()
}
